### PR TITLE
[SPARK-13930][SQL] Apply fast serialization on collect limit operator

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

JIRA: https://issues.apache.org/jira/browse/SPARK-13930

Recently the fast serialization has been introduced to collecting DataFrame/Dataset (#11664). The same technology can be used on collect limit operator too.
## How was this patch tested?

Add a benchmark for collect limit to `BenchmarkWholeStageCodegen`.

Without this patch:

```
model name      : Westmere E56xx/L56xx/X56xx (Nehalem-C)
collect limit:                      Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
-------------------------------------------------------------------------------------------
collect limit 1 million                  3413 / 3768          0.3        3255.0       1.0X
collect limit 2 millions                9728 / 10440          0.1        9277.3       0.4X
```

With this patch:

```
model name      : Westmere E56xx/L56xx/X56xx (Nehalem-C)
collect limit:                      Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
-------------------------------------------------------------------------------------------
collect limit 1 million                   833 / 1284          1.3         794.4       1.0X
collect limit 2 millions                 3348 / 4005          0.3        3193.3       0.2X
```

### CODE REVIEW

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
cc @davies @rxin

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala [hunk 1]
```diff
@@ -218,48 +218,64 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
   }
 
   /**
-   * Runs this query returning the result as an array.
+   * Packing the UnsafeRows into byte array for faster serialization.
+   * The byte arrays are in the following format:
+   * [size] [bytes of UnsafeRow] [size] [bytes of UnsafeRow] ... [-1]
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: This is the implementation details, it has nothing with the APIs, I'd like to keep these as comments.

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: nvm, you make it as an function.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala [hunk 2]
```diff
@@ -218,48 +218,64 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
...
-      var sizeOfNextRow = ins.readInt()
-      while (sizeOfNextRow >= 0) {
-        val bs = new Array[Byte](sizeOfNextRow)
-        ins.readFully(bs)
-        val row = new UnsafeRow(nFields)
-        row.pointTo(bs, sizeOfNextRow)
-        results += row
-        sizeOfNextRow = ins.readInt()
-      }
+      results ++= collectRowFromBytes(bytes)
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: Should we avoid this copy ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala [hunk 3]
```diff
@@ -218,48 +218,64 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
...
       out.close()
       Iterator(bos.toByteArray)
     }
+  }
 
-    // Collect the byte arrays back to driver, then decode them as UnsafeRows.
+  /**
+   * Collect the byte arrays back to driver, then decode them as UnsafeRows.
+   */
+  private def collectRowFromBytes(bytes: Array[Byte]): Array[InternalRow] = {
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: This function has nothong about `collect`, decodeUnsafeRow or deserializeUnsafeRows?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala (hunk 2 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: I think we should. Updated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala (hunk 3 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: done with decodeUnsafeRow.

**Comment by viirya** [comment]
cc @davies The comments are addressed and tests are passed. Please see if this is ok now. Thanks!

**Comment by davies** [comment]
LGTM, merging this into master, thanks!

**Pull request closed without merging**